lib/python.py

```diff
@@ -234,69 +173,91 @@ def separate_namespaces(tree):
-        all_vars = node_a.get_variables(call.line_number)
-
-        for var in all_vars:
-            var_match = call.matches_variable(var)
-            if var_match:
-                if var_match == 'UNKNOWN_MODULE':
-                    return None, None
-                assert isinstance(var_match, Node)
-                return var_match, None
-
-        possible_nodes = []
-        if call.is_attr():
-            for node in all_nodes:
-                if call.token == node.token and node.parent != node_a.root_parent():
-                    possible_nodes.append(node)
-        else:
-            for node in all_nodes:
-                if node.parent.group_type == 'MODULE' and call.token == node.token:
-                    possible_nodes.append(node)
-                if node.is_constructor and call.token == node.parent.token:
-                    possible_nodes.append(node)
-
-        if len(possible_nodes) == 1:
-            return possible_nodes[0], None
-        if len(possible_nodes) > 1:
-            return None, call
-        return None, None
+        token = "(global)"
+        line_number = 0
+        calls = make_calls(lines)
+        variables = make_local_variables(lines, parent)
+        return Node(token, line_number, calls, variables, parent=parent)
 
     @staticmethod
-    def make_file_group(tree, filename):
+    def make_class_group(tree, parent):
         """
-        Given an AST for the entire file, generate a file group complete with
-        subgroups, nodes, etc.
+        Given an AST for the subgroup (a class), generate that subgroup.
+        In this function, we will also need to generate all of the nodes internal
+        to the group.
 
         :param tree ast:
-        :param filename Str:
-
+        :param parent Group:
         :rtype: Group
         """
+        assert type(tree) == ast.ClassDef
         subgroup_trees, node_trees, body_trees = Python.separate_namespaces(tree)
-        group_type = 'MODULE'
-        token = os.path.split(filename)[-1].rsplit('.py', 1)[0]
-        line_number = 0
 
-        file_group = Group(token, line_number, group_type, parent=None)
+        group_type = 'CLASS'
+        token = tree.name
+        line_number = tree.lineno
+
+        class_group = Group(token, line_number, group_type, parent=parent)
 
         for node_tree in node_trees:
-            file_group.add_node(_make_node(node_tree, parent=file_group))
-        file_group.add_node(_make_root_node(body_trees, parent=file_group), is_root=True)
+            class_group.add_node(Python.make_nodes(node_tree, parent=class_group)[0])
 
         for subgroup_tree in subgroup_trees:
-            file_group.add_subgroup(_make_class_group(subgroup_tree, parent=file_group))
+            logging.warning("Code2flow does not support nested classes. Skipping %r in %r.",
+                            subgroup_tree.name, parent.token)
+        return class_group
+
+    # @staticmethod
+    # def make_file_group(tree, filename):
```

Comment:
The `make_file_group()` methods in `python.py` and `javascript.py` are moved to `engine.py`. You can delete this comment block. And the abstract method of the `BaseLanguage` class in `model.py` is not used.

Reply:
Fixed